### ISSUE TITLE
What is a const object?

### ISSUE THREAD

**Issue by rricard** [tc39/proposal-record-tuple#24]
So we're going to dump the name `const object` and favor a more different naming (for the proposal and the results of a `typeof`)

For now we have multiple ideas:

- Immutable objects/arrays - `typeof immObjOrArr === "immutable"`
- Records/Tuples - `typeof recordOrTuple=== "record"`
- _Your idea here!_

@ljharb, I am very interested in what you think!

**Comment by ljharb** [comment]
While I'm sure I'm missing some, and it's entirely possible some of these are incorrect or do not hold as absolutely as i'm implying, here's some current axioms I observe:

Objects:
 - can be used as WeakSet values, WeakMap keys, and a WeakRef target
 - `Object(x) === x` is true
 - `typeof x` is `'object'` or `'function'` (this list, of course, could grow if needed, but there's a web compat risk)
 - can have own properties; and can be frozen/sealed/preventExtensioned.

Primitives:
 - can not be used in weak contexts
 - have no properties - *all property access whatsoever* is done by creating an ephemeral boxed object, and accessing values on the prototype. `Object.keys(x)` and `Object.getOwnPropertyNames(x)` of any non-nullish primitive is always an empty array.
 - `Object(x)` creates a new boxed primitive, such that `Object(x) !== x` for any primitive `x`
 - `typeof x` is `'boolean'` or `'undefined'` or `'string'` or `'number'` or `'symbol'`, or `'bigint'`, or `'object'` for `null` (this list, of course, could grow if needed, but there's a web compat risk - adding symbols broke some code on the web, which caused it to use alternative means mentioned elsewhere in this comment)

Non-nullish instances, whether objects or primitives:
 - have a global Constructor, and a `.prototype` on that constructor
 - have at least one `Constructor.prototype.something.call(x)` method does an internal slot check on `x`, typically throwing when `x` lacks the slot (`Error` is the lone legacy exception; `Array.isArray` is a static predicate instead of an instance thrower)

It seems like this proposal seeks to violate some of these axioms.

In particular, note that primitives *do not have any own properties* - only properties shared via a prototype. If it has any own properties, *it is an object*, and must conform to the precedents set above.

So, what kind of values are produced by this proposal, taking into account the axioms

**Comment by JustFly1984** [comment]
My 2 cents: What about immutable functions?

**Comment by littledan** [comment]
@ljharb The idea of this proposal is that records and tuples are primitives. What do you see violated?

**Comment by littledan** [comment]
@JustFly1984 I don't quite understand--these are values, not functions.

**Comment by ljharb** [comment]
@littledan if that's the case, then what does `Object(record)` or `Object(tuple)` produce? As a primitive, I'd expect a boxed form of the primitive with a prototype and constructor (even tho the constructor could certainly throw, a la Symbol).

Also, if they're primitives, then `Object.keys(record)` or `Object.keys(tuple)` should be empty arrays - because primitives have no properties, only objects have properties.

In other words I think that making things with properties primitives makes no sense at all; they're either objects, or they have no properties.

**Comment by thorn0** [comment]
> Object.keys(record) or Object.keys(tuple) should be empty arrays - because primitives have no properties, only objects have properties

```console
> Object.keys('foo')
[ '0', '1', '2' ]
```
🤔

**Comment by ljharb** [comment]
@thorn0 touché :-) strings are indeed the primitive for which this isn't true - `Object.getOwnPropertyDescriptor('foo', 0)` works as well. I've updated my above comment to correct this.

In that case, my concerns about conflicting mental models are largely reduced. Given that the intention is to make these primitives, then the following criteria must apply:
 - `Object(primitive)` provides a boxed primitive
 - the boxed primitive has a global Constructor, and a .prototype on that constructor
 - there is at least one `Constructor.prototype.something.call(x)` method does an internal slot check on x, typically throwing when x lacks the slot (Error is the lone legacy exception; Array.isArray is a static predicate instead of an instance thrower)
 - the primitive can't be used in a weak context

As long as these are met, and as long as the web compat risk of adding new `typeof` values is considered, then I withdraw any objections I implied above to making them primitives.

**Comment by littledan** [comment]
The semantics I had in mind is that they would produce these boxed forms, yes. I don't think the second one holds, though--strings are primitives and have properties.

**Comment by ljharb** [comment]
@littledan you’re totally right; https://github.com/rricard/proposal-const-value-types/issues/24#issuecomment-505194756 reminded me of that, so “having properties” no longer concerns me here. My mistake.

**Comment by rdking** [comment]
@littledan Is it really that strings have properties? I always thought the strings were being converted to their boxed equivalent when properties are requested...

Anyway, I want to know what kind of immutability these `const` objects have. Is it that nothing accessed via the const object is allowed to be mutated, or is it that every property and sub-property of the const object is marked non-configurable and either non-writable or forced to have an undefined setter? I'm hoping it's not either of these. If it isn't, then what exactly **is** a const object?

**Comment by littledan** [comment]
@rdking You're right, my wording was sloppy. Strings don't have properties; these are created when ToObject is implicitly called.

I'd suggest that records and tuples work analogously: that they create objects whose prototype is `Record.prototype` or `Tuple.prototype`, which has an internal slot with the underlying record, and which are exotic objects, where property access gets the record or tuple slot, and which generally act like they are frozen in response to meta-object operations.

I believe @mheiber may be in the process of writing this description up a bit more thoroughly in the explainer.

**Comment by rdking** [comment]
Here's an idea. Why not have a simple base class (maybe called `Immutable`), and let that be the base for all such immutable objects. This way, if someone wants to create their own immutables, they can simply extend this base. I don't know how the semantics for this will work, especially in the face of fields. You'd have to somehow allow the instance to remain mutable until all constructors have completed.

**Comment by littledan** [comment]
@rdking I don't know how the semantics for that would work either. This proposal is strongly based around objects which are immutable in a consistent way, across their lifetime, given that we don't have the freedom to check in the compiler that there's a single assignment to the field in the constructor using data flow analysis the way other programming languages do. We have to consider the object leaking at various different points, and maintain consistency in all of that.

**Comment by rdking** [comment]
@littledan Took me a little bit of time to come up with something. You might like it. Just as there is a `new.target` inside constructor functions, let there also be a boolean internal property `[[Immutable]]`. Immediately following the completion of  `new.target`, but before the result is returned to the caller, flag the object as immutable. The engine would then need to perform 1 extra check for every set operation: if the receiver is immutable, throw.

**Comment by rricard** [comment]
We clarified the vocabulary on the proposal and we presented the record and tuple boxing during the Stage 1 presentation. Now the issue seems stale, please feel free to ping if you think we did not get to the bottom of the issue.
